entities/sync/sync_test.go

```diff
@@ -70,6 +81,185 @@ func TestKeyLockerLockUnlock(t *testing.T) {
+// all should give up after cancellation and return false from TryLockWithContext
+func TestKeyLockerContextMutexLockConcurrentCancel(t *testing.T) {
+	s := NewKeyLockerContext()
+	s.Lock("t1")
+	defer s.Unlock("t1")
+
+	numGoroutines := 10
+	ctx, cancel := context.WithCancel(context.Background())
+	wgOuter := sync.WaitGroup{}
+	wgOuter.Add(numGoroutines)
+	wgInner := sync.WaitGroup{}
+	wgInner.Add(numGoroutines)
+	counter := atomic.Int32{}
+	// try to lock concurrently, should all wait
+	for i := 0; i < numGoroutines; i++ {
+		go func() {
+			done := s.TryLockWithContext("t1", ctx)
+			if !done {
+				counter.Add(1)
+			}
+			wgInner.Done()
+		}()
+		wgOuter.Done()
+	}
+	wgOuter.Wait() // make sure all goroutines are started
+	cancel()       // cancel context to stop trying to lock
+
+	// now all goroutines should have given up getting the lock
+	wgInner.Wait()
+	require.Equal(t, int32(numGoroutines), counter.Load())
+}
+
+func TestKeyLockerContextMutexLockConcurrentUnlock(t *testing.T) {
+	s := NewKeyLockerContext()
+	s.Lock("t1")
+
+	numGoroutines := 10
+	wgOuter := sync.WaitGroup{}
+	wgOuter.Add(numGoroutines)
+	wgInner := sync.WaitGroup{}
+	wgInner.Add(numGoroutines)
+	counter := atomic.Int32{}
+	// try to lock concurrently, should all wait
+	for i := 0; i < numGoroutines; i++ {
+		go func() {
+			done := s.TryLockWithContext("t1", context.Background())
+			if !done {
+				counter.Add(1)
+			} else {
+				counter.Add(-1)
+				s.Unlock("t1")
+			}
+			wgInner.Done()
+		}()
+		wgOuter.Done()
+	}
+	wgOuter.Wait() // make sure all goroutines are started
+	s.Unlock("t1") // unlock so that one of the goroutines can acquire the lock
+
+	wgInner.Wait() // wait for all goroutines to be done
+	require.Equal(t, -int32(numGoroutines), counter.Load())
+}
+
+func TestKeyLockerContextMultipleContext(t *testing.T) {
+	s := NewKeyLockerContext()
+	s.Lock("t1")
+
+	numGoroutines := 10
+
+	contexts := make([]struct {
+		context context.Context
+		cancel1 context.CancelFunc
+	}, numGoroutines)
+	for i := range contexts {
+		contexts[i].context, contexts[i].cancel1 = context.WithCancel(context.Background())
+	}
+
+	wgOuter := sync.WaitGroup{}
+	wgOuter.Add(numGoroutines)
+	wgInner := sync.WaitGroup{}
+	wgInner.Add(numGoroutines)
+	counterCancelled := atomic.Int32{}
+	counterSucceded := atomic.Int32{}
+	// try to lock concurrently, should all wait
+	for i := 0; i < numGoroutines; i++ {
+		go func() {
+			done := s.TryLockWithContext("t1", contexts[i].context)
+			if !done {
+				counterCancelled.Add(1)
+			} else {
+				counterSucceded.Add(1)
+				s.Unlock("t1")
+			}
+			wgInner.Done()
+		}()
+		wgOuter.Done()
+	}
+	wgOuter.Wait() // make sure all goroutines are started
+
+	// cancel some contexts
+	for i := range contexts[:5] {
+		contexts[i].cancel1()
+	}
+	time.Sleep(100 * time.Millisecond) // wait a little to give goroutines time to react to cancellation
```

Comment:
can be source of flakiness. maybe introduce separate waitgroup=5 for cancelled trylock calls?